src/runtime_src/core/tools/xbtracer/src/lib/logger.cpp

```diff
@@ -0,0 +1,334 @@
+  pid = GetCurrentProcessId();
+
+  TCHAR buffer[str_sz_l];
+  DWORD result =
+          GetEnvironmentVariable(TEXT("INST_DEBUG"), buffer, str_sz_l);
+  if (result > 0 && result < str_sz_l && !strcmp(buffer, "TRUE"))
+    inst_debug=true;
+
+  result = GetEnvironmentVariable(TEXT("TRACE_APP_NAME"), buffer, str_sz_l);
+  if (result > 0 && result < str_sz_l)
+    program_name = std::string(buffer);
+#else
+
+  std::lock_guard<std::mutex> lock(env_mutex);
+
+  //NOLINTNEXTLINE(concurrency-mt-unsafe) - protected by env_mutex
+  char* app_dbg_env = getenv("INST_DEBUG");
+  if ( app_dbg_env ) {
+    if(!strcmp(app_dbg_env, "TRUE")) {
+      inst_debug = true;
+    }
+  }
+
+  //NOLINTNEXTLINE(concurrency-mt-unsafe) - protected by env_mutex
+  program_name=getenv("TRACE_APP_NAME");
+
+  pid = getpid();
+
+  // Retrieve the time from the environment variable
+  //NOLINTNEXTLINE(concurrency-mt-unsafe) - protected by env_mutex
+  const char* time_str = getenv("START_TIME");
+  if (time_str == nullptr)
+    std::cerr << "Environment variable START_TIME not set!" << std::endl;
+  else
+    std::cout << "Time retrieved from environment variable: " << time_str
+              << std::endl;
+
+  // Convert the string back to a timespec structure
+  std::istringstream iss(time_str);
+  iss >> start_time.tv_sec;
+  iss.ignore(1, '.');  // Ignore the dot character
+  iss >> start_time.tv_nsec;
+#endif /* #ifdef _WIN32 */
+
+  read_time_now_fmt(time_fmt_str, start_time);
+
+  namespace fs = std::filesystem;
+  // Create directory using the timestamp.
+  namespace fs = std::filesystem;
+  if (!fs::exists(time_fmt_str))
+    if (!fs::create_directory(time_fmt_str))
+      std::cerr << "Failed to create directory: " << time_fmt_str << std::endl;
+
+  // Construct full path and open files for logging.
+  oss_full_path << "." <<path_separator << time_fmt_str << path_separator
+                << xrt_trace_filename;
+
+  fp.open(oss_full_path.str(), std::ios::out);
+
+  oss_full_path.str("");
+  oss_full_path.clear();
+
+  oss_full_path << "." << path_separator << time_fmt_str << path_separator
+                << xrt_trace_bin_filename;
+
+  fp_bin.open(oss_full_path.str(), std::ios::out | std::ios::binary);
+
+  fp << "|HEADER|pname:\"" << program_name <<  "\"|pid:" << pid << "|xrt_ver:"
+     << XRT_DRIVER_VERSION << "|os:" << os_name_ver() << "|time:"
+     << time_fmt_str << "." << std::setfill('0') << std::setw(fw_9)
+     << start_time.tv_nsec << "|\n";
+
+  fp << "|START|"<< time_fmt_str << "." << std::setfill('0') << std::setw(fw_9)\
+     << start_time.tv_nsec << "|\n";
+}
+
+/*
+ * destructor
+ * */
+logger::~logger()
+{
+  std::string time_fmt_str="";
+  struct timespec time_now{};
+
+  #ifndef _WIN32
+  clock_gettime (CLOCK_REALTIME, &time_now);
+  #endif
+
+  read_time_now_fmt(time_fmt_str, time_now);
+
+  fp << "|END|" << time_fmt_str << "." << std::setfill('0') << std::setw(fw_9)
+     << time_now.tv_nsec << "|\n";
+
+  fp_bin.close();
+  fp.close();
```

Comment:
Almost every single line in this dtor can throw.   Dtors are now allowed to throw, do try/catch .

Reply:
Wil do it in next update.